### PR TITLE
Fix copy/paste bug in EditingService.cs

### PR DESCRIPTION
**Pull request by Zofware**
Fix a copy/paste bug in EditingService.cs that caused the data grid view state to go haywire when using inline editing mode on rows with MVVM properties (i.e. properties that send PropertyChanged notifications).

Also, ensure the row is refreshed/redrawn when inline edits are cancelled. Previously, canceled edits would still appear on the row until the view was refreshed/redrawn by other means such as resizing the column.

### CODE REVIEW

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs [hunk 1]
```diff
@@ -85,7 +85,7 @@ public bool BeginEdit(ItemInfo rowInfo)
                         this.Owner.EditRowLayer.ScheduleFirstEditorForFocus();
                     }
 
-                    this.Owner.Model.CurrentDataProvider.CommitEditOperation(this.EditItem);
+                    this.Owner.Model.CurrentDataProvider.BeginEditOperation(this.EditItem);
```
Comment (Zofware) [inline Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:88]: Fix typo, probably resulting from a copy/paste error.

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs [hunk 2]
```diff
@@ -181,27 +181,29 @@ public bool CancelEdit(ActionTrigger trigger)
...
                 }
                 else
                 {
                     this.Owner.Model.CancelEdit();
                     this.Owner.EditRowLayer.EditorLayoutSlots.Clear();
+                    this.Owner.CurrencyService.RefreshCurrentItem(true);
                 }
 
-                this.Owner.Model.CurrentDataProvider.CommitEditOperation(this.EditItem);
+                this.Owner.Model.CurrentDataProvider.CancelEditOperation(this.EditItem);
```
Comment (Zofware) [inline Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:194]: Fix typo, probably resulting from a copy/paste error.

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs [hunk 3]
```diff
@@ -181,27 +181,29 @@ public bool CancelEdit(ActionTrigger trigger)
...
             };
 
-            this.Owner.updateService.RegisterUpdate(update);
-
             foreach (var pair in this.operation.OriginalValues)
             {
                 pair.Key.SetValueForInstance(this.operation.EditItemInfo.Item, pair.Value);
             }
 
+            this.Owner.updateService.RegisterUpdate(update);
```
Comment (Zofware) [inline Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:205]: Register the update after restoring the row's original values to ensure they are shown onscreen.

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs [hunk 4]
```diff
@@ -181,27 +181,29 @@ public bool CancelEdit(ActionTrigger trigger)
...
                     this.Owner.EditRowLayer.EditorLayoutSlots.Clear();
+                    this.Owner.CurrencyService.RefreshCurrentItem(true);
                 }
 
-                this.Owner.Model.CurrentDataProvider.CommitEditOperation(this.EditItem);
+                this.Owner.Model.CurrentDataProvider.CancelEditOperation(this.EditItem);
             })
             {
-                Flags = UpdateFlags.AffectsContent
+                Flags = UpdateFlags.AffectsData
```
Comment (Zofware) [inline Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:197]: This ensures the row is redrawn.

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs [hunk 5]
```diff
@@ -181,27 +181,29 @@ public bool CancelEdit(ActionTrigger trigger)
...
                     {
                         this.Owner.ExternalEditor.CancelEdit();
+                        this.Owner.CurrencyService.RefreshCurrentItem(true);
                     }
                 }
                 else
                 {
                     this.Owner.Model.CancelEdit();
                     this.Owner.EditRowLayer.EditorLayoutSlots.Clear();
+                    this.Owner.CurrencyService.RefreshCurrentItem(true);
```
Comment (Zofware) [inline Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:191]: Ensure the row is redrawn.

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs [hunk 6]
```diff
@@ -181,27 +181,29 @@ public bool CancelEdit(ActionTrigger trigger)
                     if (trigger != ActionTrigger.ExternalEditor)
                     {
                         this.Owner.ExternalEditor.CancelEdit();
+                        this.Owner.CurrencyService.RefreshCurrentItem(true);
```
Comment (Zofware) [inline Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:184]: Ensure the row is redrawn.

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs [hunk 7]
```diff
@@ -118,18 +118,20 @@ public bool CommitEdit(ActionTrigger trigger)
...
                         this.Owner.Model.CommitEdit();
                         this.Owner.EditRowLayer.EditorLayoutSlots.Clear();
+                        this.Owner.CurrencyService.RefreshCurrentItem(true);
                     }
 
                     this.Owner.Model.CurrentDataProvider.CommitEditOperation(this.EditItem);
                 })
                 {
-                    Flags = UpdateFlags.AffectsContent,
+                    Flags = UpdateFlags.AffectsData,
```
Comment (vvgeorgiev) [inline Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:134]: Why is this change needed? AffectsData has a lower priority compared to AffectsContent and this would affect how grid is rebuilt.

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs [hunk 8]
```diff
@@ -118,18 +118,20 @@ public bool CommitEdit(ActionTrigger trigger)
                         if (trigger != ActionTrigger.ExternalEditor)
                         {
                             this.Owner.ExternalEditor.CommitEdit();
+                            this.Owner.CurrencyService.RefreshCurrentItem(true);
```
Comment (vvgeorgiev) [inline Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:121]: Why is this needed? Did you find any issue with current item scrolling out of view?

**Comment by Zofware** [comment]
Hello @vvgeorgiev, 

Thanks for your comments. I created a very [simple UWP sample app](https://github.com/Zofware/EditingServiceBug) that demonstrates the initial bug I was trying to fix. Specifically, that RadDataGrid cells backed by MVVM properties cannot be displayed or edited. Let's use that as a baseline demo for this PR.

The next thing I'll need to do is to somehow demonstrate that fixing the probable copy/paste typo in the EditingService code at least allows editing cells backed by MVVM properties. I probably need to create a build of Telerik UI for UWP from my branch for this.

Following that, I'll attempt to justify why the additional changes are needed. (Or convince myself that they are not! 😄)

In any event, the initial sample showing the bug is all I have time for at the moment, but I'll revisit this when I can.

Thanks,
Adam

**Comment by Zofware** [comment]
@vvgeorgiev,

Thanks for posting your feedback. I sent a reply a few hours ago that was incorrect so I deleted it. It had a link to a sample app that was supposed to demonstrate the issue I'm trying to fix, but the sample itself was buggy. Once I straighten out the sample, I'll repost and we can continue the discussion.

Thanks,
Adam

**Comment by Zofware** [comment]
@vvgeorgiev,

Here's a [short YouTube video](https://youtu.be/mvAG5vz9Wpw) showing why I think at least some additional refresh logic is needed after fixing the two one-line copy/paste bugs in the first commit of this PR. The problem is when edits are canceled, the original values are not shown on the affected row until something else causes a redraw (like another edit operation on the same row).

It's possible that moving the `RegisterUpdate` call, as in the 2nd commit of this PR, is all that's needed to get the row to redraw correctly after canceling an edit but I don't think so. I'll have to check again since it's been a few months since I made these changes.

Regardless, some additional refresh logic is required after canceling an edit. Please suggest a more appropriate way of doing this if you know it. My familiarity with this code is pretty limited.

Thanks,
Adam

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs (hunk 7 above)
Comment (Zofware) [reply Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:134]: See https://github.com/telerik/UI-For-UWP/pull/513#issuecomment-1283460780

File: Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs (hunk 8 above)
Comment (Zofware) [reply Controls/Grid/Grid.UWP/View/Services/Editing/EditingService.cs:121]: See https://github.com/telerik/UI-For-UWP/pull/513#issuecomment-1283460780

**Comment by vvgeorgiev** [comment]
Hi Adam, 
Thanks for the additional details.

Your initial commit is no doubt correct. Moving the `RegisterUpdate` wouldn't matter much runtime, as it's executed asynchronously. Still, it's correct. 
Now, I seem to struggle to reproduce the problem as seen in your video. Sharing that app, you used for the video might be a good idea.
Thanks.
Ves

**Comment by vvgeorgiev** [comment]
Hi Adam, 
Thanks for the app. I am afraid it does not reproduce the issue on my end. I am not sure if the Windows version matters, or there's anything else. My suggestion here would be to go ahead and merge the core update i.e. the first commit, along with [this change](https://github.com/telerik/UI-For-UWP/pull/513/commits/773ddf8fc1987ca48667938546ef0bcb7513e285#diff-a0fd49d3de630c367b2eac1f5152ea15564f870926a076037c06ac6e25c62b99R205), while further investigation is needed on the side effects you have shown.
Thanks,
Ves

**Comment by Zofware** [comment]
FWIW, I'm running Windows 10 22H2 (19045.2130).

**Comment by Zofware** [comment]
I'll roll back the last two commits, leaving only the first two. Thanks.

**Comment by Zofware** [comment]
Last two commits reverted.

**Comment by vvgeorgiev** [comment]
I am sorry, I did not make it clear. I suggest that we go ahead with the first commit and only a single change from the second one -- the one that moves the call to UpdateService.RegisterUpdate from line 198 to line 205.

**Comment by Zofware** [comment]
Sorry for the misunderstanding. Fixed.

**Review by vvgeorgiev — APPROVED** [review]
